modules/base/src/algorithm/mesh/meshclipping.cpp

```diff
@@ -0,0 +1,609 @@
+/*********************************************************************************
+ *
+ * Inviwo - Interactive Visualization Workshop
+ *
+ * Copyright (c) 2018-2019 Inviwo Foundation
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+#include <modules/base/algorithm/mesh/meshclipping.h>
+#include <inviwo/core/datastructures/geometry/simplemesh.h>
+#include <inviwo/core/datastructures/geometry/basicmesh.h>
+#include <inviwo/core/datastructures/geometry/edge.h>
+#include <inviwo/core/datastructures/geometry/polygon.h>
+
+namespace inviwo {
+
+namespace meshutil {
+
+namespace detail {
+
+// Check point equality with threshold
+inline bool equal(vec3 v1, vec3 v2, float eps) {
+    return (std::fabs(v1.x - v2.x) < eps && std::fabs(v1.y - v2.y) < eps &&
```

Comment:
I tested it and unfortunately the epsilon needs a bit of custom tuning, meaning when we use ULP or machine epsilon, `findLoops` does not find connection points anymore. The current value of `const float EPSILON = 0.00001f` works fine for my tests. I know, this is not optimal and assuming that the intersection points (triangle edges with plane) are exact, machine epsilon **should** work. Please let me know, if any of your tests fail. Otherwise, keep it this way?

Reply:
you could use a larger ULP for almostEqual. Otherwise I think the epsilon need to either be derived from the size of the mesh. i.e. norm(mesh bounding box)/10^6 or something like that.
a hardcoded EPSILON like 0.000001 will only work for meshes that has a bounding box of size ~1. But we can often have much larger ~10000 or much smaller ~0.0001 meshes...
I you can derive it from the mesh. then It should at least be possible for the user to supply a epsilon....

Reply:
I implemented the mesh-dependent epsilon, but one case where it still fails are thin "slab" meshes. You can reproduce this in the [Large Epsilon Regression Test](https://github.com/inviwo/inviwo/blob/feature/mesh-plane-clipping/modules/base/tests/regression/meshclipping-large-epsilon/meshclipping-large-epsilon.inv), making a thin "cube" with the Mesh Creator.

It throws [this Exception](https://github.com/inviwo/inviwo/blob/feature/mesh-plane-clipping/modules/base/src/algorithm/mesh/meshclipping.cpp#L144) in findLoops(), saying some edges are fully unconnected. This needs some investigation...